flow360/component/simulation/translator/utils.py

```diff
@@ -136,29 +132,70 @@ def remove_units_in_dict(input_dict):
     return input_dict
 
 
+def is_terminal_instance(obj, cls):
+    """Check if an object is an instance of a class and not a subclass."""
+    if not isinstance(obj, cls):
+        return False
+    # Check if there are any subclasses of cls
+    subclasses = cls.__subclasses__()
+    for subclass in subclasses:
+        if isinstance(obj, subclass):
+            return False
+    return True
+
+
 def has_instance_in_list(obj_list: list, class_type):
     """Check if a list contains an instance of a given type."""
     if obj_list is not None:
         for obj in obj_list:
-            if isinstance(obj, class_type):
+            if is_terminal_instance(obj, class_type):
                 return True
     return False
 
 
+def _is_last_of_type(lst, obj):
+    current_type = type(obj)
+    last_index = -1
+
+    for i, item in enumerate(lst):
+        if is_terminal_instance(item, current_type):
+            last_index = i
+
+    if last_index == -1:
+        return False  # The type of obj does not exist in the list
+
+    return lst[last_index] == obj
+
+
 def get_attribute_from_first_instance(
-    obj_list: list, class_type, attribute_name: str, check_empty_entities: bool = False
+    obj_list: list, class_type, attribute_name: str, use_empty_entities: bool = False
 ):
     """In a list loop and find the first instance matching the given type and retrive the attribute"""
     if obj_list is not None:
         for obj in obj_list:
-            if isinstance(obj, class_type):
-                if check_empty_entities and obj.entities is not None:
+            if (
+                is_terminal_instance(obj, class_type)
+                and getattr(obj, attribute_name, None) is not None
+            ):
+                # Route 1: Requested to look into empty-entity instances
+                if use_empty_entities and getattr(obj, "entities", None) is not None:
+                    # We only look for empty entities instances
+                    # Note: This poses requirement that entity list has to be under attribute name 'entities'
+                    continue
+
+                # Route 2: Allowed to look into non-empty-entity instances
+                default_value = obj.model_fields[attribute_name].default
+                field_value = getattr(obj, attribute_name)
+                is_last_item = _is_last_of_type(obj_list, obj)
+
+                if (field_value == default_value) and (is_last_item is False):
```

Comment:
What's the current design of validation if two objects have different values of a global entry? Shouldn't this fail validation?

Reply:
To be added in solver capability validators